Added proposed implementation of RawRepresentable extension and corresponding test file.

SamHastings1066 opened the pull request:
This change resolves #32 by adding the proposed implementation of the RawRepresentable extension and a corresponding test file.

Please let me know if there is anything else you want me to do before I mark this ready for review.

I have two observations that I wanted to highlight and get your feedback on:

1) I implemented the extension using the for unwrap (`!`) in `self.init(rawvalue: RawValue(storage))!` as was originally proposed. This assumes the RawValue initializer will never fail. Is this a fair assumption given its intended use in this project?
2) I included a test in the BitFieldProjectableTests.swift file called `testInitializationFomInvalidValue` to highlight the fact that Swift is not enforcing a bit width constraint at the enum level. This means a user could define an enum case that doesn't adhere to the `bitWidth` constraint without encountering a compile-time error. I wanted to make sure that this is OK. If so I can remove the test and the corresponding `invalidValue` case.

rauhul commented:
I did a quick look over the code (and will think more about this tomorrow morning) and I am not 100% sure about `self.init(rawValue: RawValue(storage))!` as you also mention in the PR body.

> I implemented the extension using the for unwrap (!) in self.init(rawvalue: RawValue(storage))! as was originally proposed. This assumes the RawValue initializer will never fail. Is this a fair assumption given its intended use in this project?

I'm wondering if we should leave out the default `init`  implementation.

At the very least we should replace the force unwrap with a guard + precondition failure describing the error

rauhul commented on .swiftpm/xcode/xcshareddata/xcschemes/swift-mmio.xcscheme:
```diff
@@ -0,0 +1,140 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can we drop this file from the PR?

rauhul commented on Tests/MMIOTests/BitFieldProjectableTests.swift:
```diff
@@ -0,0 +1,93 @@
+    case master_spi = 0x3
+    case invalidValue = 0x4
+    
+    public static var bitWidth: Int { 2 }
```
Could we make these bitWidths `fileprivate` as well

rauhul commented on Sources/MMIO/BitFieldProjectable.swift:
```diff
@@ -43,6 +43,16 @@ extension Bool: BitFieldProjectable {
 
+extension RawRepresentable where Self: BitFieldProjectable, RawValue: FixedWidthInteger {
+  public init<Storage>(storage: Storage) where Storage: FixedWidthInteger & UnsignedInteger {
+      self.init(rawValue: RawValue(storage))!
```
Definitely need to use a precondition failure with a message explaining the issue

rauhul commented on Sources/MMIO/BitFieldProjectable.swift:
```diff
@@ -43,6 +43,16 @@ extension Bool: BitFieldProjectable {
+  }
+
+  public func storage<Storage>(_: Storage.Type) -> Storage where Storage: FixedWidthInteger & UnsignedInteger {
+    Storage(rawValue)
```
I think this need a precondition `MemoryLayout<Storage>.size * 8  >= self.bitWidth`

rauhul reviewed (changes requested):
Some notes, also please run `make lint` and `make format`

SamHastings1066 commented:
I refined the RawRepresentable extension with the proposed safety checks and minor fixes:
 
- Added precondition failures to ensure initialization fails when the storage value is invalid and to ensure storage size is sufficient.
- Added documentation-only tests to validate the new precondition failures.
- Removed unnecessary scheme file.
- Changed bitWidth access level from public to fileprivate.
- Performed code formatting and linting.

rauhul commented:
Hi! I'm back from vacation, the last couple things I'd like to get changed before merging is making these tests into file check based tests:
```
  /// Tests that the initializer triggers a precondition failure when provided with an invalid storage value.
  func testInitializationFromInvalidStorageValue()

  /// Checks that a precondition failure is triggered when the storage type does not have enough bits to represent the bitWidth of the value.
  func testStoragePreconditionFailureForInsufficientBitWidth()
```

SamHastings1066 commented:
Hi Rauhul, now it's me who's back from vacation! Before I attempt an implementation of those two tests as file check based tests, can I verify that the approach proposed below is correct:

1. Create a directory named PreconditionFailureTests within the MMIOFileCheckTests directory. 
2. Place the error-generating code from the current precondition failure tests into two new Swift files in this directory.
3. Add a new method to the pre-existing class MMIOFileCheckTests to run the precondition failure tests. This method will compile the error generating code and store the output in temp files, validate the contents of these temp files against the expected patterns using FileCheck, remove the temp files.

I want to make sure that this is what you're after because I think it will be a bit involved so I don't want to go down a rabbit hole if it's not necessary.

rauhul commented:
> Hi Rauhul, now it's me who's back from vacation! Before I attempt an implementation of those two tests as file check based tests, can I verify that the approach proposed below is correct:
> 
> 1. Create a directory named PreconditionFailureTests within the MMIOFileCheckTests directory.
> 2. Place the error-generating code from the current precondition failure tests into two new Swift files in this directory.
> 3. Add a new method to the pre-existing class MMIOFileCheckTests to run the precondition failure tests. This method will compile the error generating code and store the output in temp files, validate the contents of these temp files against the expected patterns using FileCheck, remove the temp files.
> 
> I want to make sure that this is what you're after because I think it will be a bit involved so I don't want to go down a rabbit hole if it's not necessary.

Welcome back! I think there's a misunderstanding about the file check tests, what these do is compile test files located under `MMIOFileCheckTests/Tests` into llvm IR and then check that the IR has the expected content. This will allow us to check that invalid uses of RawRepresentable result in a trap.

See https://github.com/apple/swift-mmio/blob/main/Tests/MMIOFileCheckTests/Tests/TestRegisterModifyProjectedMismatchedWidthIsStaticTrap.swift as an example

SamHastings1066 commented:
Hi Rahul,

Apologies, I'm struggling with this - it's my first time using FileCheck.

My understanding is that FileCheck is typically used to verify the output of compiled code, such as LLVM IR, rather than runtime console output. The test you referred to, along with the other tests in the Tests folder, seem to use FileCheck in this way - to examine the LLVM IR generated by the Swift compiler. The precondition failure that my test is intended to generate occurs at runtime, and to use FileCheck for verifying the contents of this runtime error, I would need to capture it into stdout or a file that FileCheck can then read from. I don't think I can deliver this simply by adding another file to the Tests directory and then running the existing testing infrastructure, since it's not designed to handle runtime outputs.

Is my understanding is correct? Have you got any more advice on how to proceed with this test using FileCheck, or if there's an alternative approach for verifying runtime errors?

rauhul commented:
All good, any tool has a learning curve.

The `TestRegisterModifyProjectedMismatchedWidthIsStaticTrap` test is a good example because it statically verifies a runtime trap.

The gist of the test is:
- we have a register with a 2 bit field projected as a single bit type: `@ReadWrite(bits: 0..<2, as: Bool.self)`
- we perform a read-modify-write of the register `r8.modify { $0.lo = true }` using this mismatched bit-width
- we assert that the optimized LLVM IR constant folds down to a trap `// CHECK: call void @llvm.trap()`

This pattern allows us to test runtime trap behavior statically. Does that help clarify the test?

SamHastings1066 commented:
Hi Rahul,

Thank you for the clarification. I understand the approach used in TestRegisterModifyProjectedMismatchedWidthIsStaticTrap where a compile-time detectable condition is optimized into an immediate llvm.trap() call in the LLVM IR. This makes sense for scenarios where the compiler can statically determine that a trap must occur (e.g. mismatched bit widths).

However, in my test cases, the precondition failures are triggered by a runtime condition (e.g. 0xFF not mapping to any case in the UMSEL0Values enum). I don't think these types of conditions are things the Swift compiler can detect at compile time to optimize into a trap in the IR. They seem inherently tied to runtime behavior.

Could you advise on whether there's a way to adapt this test to fit the static analysis approach, or if we need to consider a different testing strategy for this particular case?

rauhul commented:
superseded by #81
